Build-process.md

```diff
@@ -0,0 +1,114 @@
+
+* Dev mode
+* Prod mode
+  * Maintenance mode
+  * Deploy mode
+  * Minify only third-party libraries
+
+Note that maintenance and deploy mode can be enabled simultaneously.
+
+We use prod mode for most everything besides local development. For example, we use prod mode when deploying to production and testing servers, and we run most tests in prod mode. A notable exception is the lighthouse accessibility tests, which run in dev mode because they don't depend on prod mode, and dev mode compilation is faster. Note that we also sometimes use prod mode locally when debugging to run a version of the app that's closer to the deployed version.
+
+
+### Build mode constants
+
+The build mode affects the application during runtimes through the following constants in `assets/constants.ts`:
+
+* `DEV_MODE`: Set to `true` if and only if we are in dev mode. Otherwise, this is `false`.
+* `EMULATOR_MODE`: Set to `true` if and only if we are not in deploy mode. Note that being in deploy mode implies being in prod mode. Otherwise, this is `false`.
+* `ENABLE_MAINTENANCE_MODE`: Set to `true` if and only if we are in maintenance mode. Note that being in maintenance mode implies being in prod mode. Otherwise, this is `false`.
+
+### Dev mode
+
+We use dev mode for local development. For example, when you run `python -m scripts.start`, you build the app in dev mode. This mode is optimized to build as quickly as possible, so it doesn't do much optimization.
+
+Here's what happens during a dev mode build:
+
+1. Some third-party libraries are generated. If you recently ran the server, then there might not be much to change here. Otherwise, the third-party libraries from `manifest.json` are downloaded. Note that the libraries in `manifest.json` are just some of our frontend libraries. The downloaded CSS and JavaScript code files are combined into `third_party/generated/third_party.css` and `third_party/generated/third_party.js`, and the downloaded fonts are placed in `third_party/webfonts/`.
+2. The build script sets the constants in `assets/constants.ts` to enable dev mode. Specifically, `DEV_MODE` is `true`, `EMULATOR_MODE` is `true`, and the other constants are `false`.
+3. [Webpack](https://webpack.js.org/) bundles our frontend code files and runs with a [dev configuration](https://github.com/oppia/oppia/blob/develop/webpack.dev.config.ts) designed for minimal compilation time. Note that the dev mode configuration file inherits from the [root webpack config file](https://github.com/oppia/oppia/blob/develop/webpack.common.config.ts). In dev mode, webpack runs in "watch" mode so that when files are changed, webpack re-builds the app automatically. Note that **this step is not handled by `build.py`.** Instead, this step is usually performed by whatever script you actually executed, for example `start.py`. When running the frontend tests, this step is triggered by configuration in `karma.conf.ts`.
```

Comment:
Maybe worth explaining, conceptually, why this step is not handled by build.py. (Or should it be?)

This is so that devs who edit build.py have a conceptual understanding of what's in scope for that file and what isn't.

Reply:
I think it should be handled by `build.py`, but I've reached out to @vojtechjelinek to check. If they agree, I'll open an issue

Reply:
If we are talking about FE tests then it is because it is required to be inside karma.

For the start.py it is because the webpack is in watch mode and if this was done in build.py we would have to keep the build part running so that we can keep the watch mode.

Reply:
I've updated the page to explain this